optaplanner-docs/src/main/asciidoc/ConstraintStreams/ConstraintStreams-chapter.adoc

```diff
@@ -58,55 +45,283 @@ that uses a functional approach:
+incremental score calculation.
+This is an example of the same code, using the Constraint Streams API:
 
-On the other hand, the ConstraintStreams API uses <<incrementalScoreCalculation,deltas>>:
+// TODO Rewrite the above paragraph PLANNER-1709 is fixed, mentioning constraint justification as another benefit.
 
-////
-TODO Replace the previous sentence when PLANNER-1709 is fixed.
+[source,java,options="nowrap"]
+----
+    private Constraint doNotAssignAnn(ConstraintFactory factory) {
+        return factory.from(Shift.class)
+                .filter(Shift::isEmployeeAnn)
+                .penalize("Don't assign Ann", HardSoftScore.ONE_SOFT);
+    }
+----
 
-So it doesn't do <<incrementalScoreCalculation,deltas>>.
-Nor does it support <<explainingTheScore,justifications>>.
+This constraint stream goes over all instances of class `Shift` amongst <<problemFacts,problem facts>> and
+<<planningEntity,planning entities>> in the <<planningProblemAndPlanningSolution,planning problem>>.
+It finds all which are assigned to employee `Ann` and for every such instance (also called "a match"), it adds a soft
+penalty of 1 to the overall <<calculateTheScore,score>>.
+The following figure illustrates this process on a problem with 4 different shifts:
+
+image::ConstraintStreams/constraintStreamIntroduction.png[align="center"]
+
+If any of the instances change during solving, the constraint stream automatically detects it and only recalculates
+the minimum necessary portion of the problem that is affected by the change.
+The following figure illustrates this <<incrementalScoreCalculation,incremental score calculation>>:
+
+image::ConstraintStreams/constraintStreamIncrementalCalculation.png[align="center"]
 
-On the other hand, the ConstraintStreams API uses <<incrementalScoreCalculation,deltas>>
-and supports <<explainingTheScore,justifications>>:
 ////
+TODO Include and describe the following image when PLANNER-1709 is fixed.
+image::ConstraintStreams/constraintStreamJustification.png[align="center"]
+////
+
+
+[[constraintStreamsFirst]]
+== Your First Constraint Stream
+
+To use the ConstraintStreams API in your project, first write a pure Java `ConstraintProvider` implementation like this:
 
 [source,java,options="nowrap"]
 ----
-    private Constraint speakerUnavailableTimeslot(ConstraintFactory factory) {
-        return factory.from(Talk.class)
-                .filter(Talk::hasAnyUnavailableSpeaker)
-                .penalize("Speaker unavailable timeslot",
-                        HardSoftScore.ONE_SOFT,
-                        Talk::getDurationInMinutes);
+    public class MyConstraintProvider implements ConstraintProvider {
+
+        @Override
+        public Constraint[] defineConstraints(ConstraintFactory factory) {
+            return new Constraint[] { // Provide as many constraints as you like. This example only provides one.
+                    firstConstraint(factory)
```

Comment:
Important: No foo's, no bar's, no firstConstraints, secondConstraints, etc.
It's for readers often hard to understand what it does.

Copy paste something from a real example - preferably an self-explaining one - and simplify it as much as we can without hurting the truth. For example, take a constraint from employee rostering, course scheduling or cloud balancing.